geode-core/src/upgradeTest/java/org/apache/geode/security/AuthExpirationFunctionDUnitTest.java

```diff
@@ -0,0 +1,142 @@
+        "org.apache.geode.management.internal.security.TestFunctions*");
+    serverVM = lsRule.startServerVM(0, properties);
+
+    serverVM.invoke(() -> {
+      Objects.requireNonNull(ClusterStartupRule.getCache())
+          .createRegionFactory(RegionShortcut.REPLICATE).create("region");
+    });
+    int serverPort = serverVM.getPort();
+    currentVersionClientVM = lsRule.startClientVM(1, CURRENT_VERSION, c1 -> c1
+        .withProperty(SECURITY_CLIENT_AUTH_INIT, UpdatableUserAuthInitialize.class.getName())
+        .withPoolSubscription(true)
+        .withServerConnection(serverPort));
+    releaseVersionClientVM = lsRule.startClientVM(2, RELEASE_VERSION,
+        c -> c.withProperty(SECURITY_CLIENT_AUTH_INIT, UpdatableUserAuthInitialize.class.getName())
+            .withPoolSubscription(true)
+            .withServerConnection(serverPort));
+
+    VMProvider.invokeInEveryMember(() -> writeFunction = new TestFunctions.WriteFunction(),
+        serverVM, currentVersionClientVM, releaseVersionClientVM);
+  }
+
+  @After
+  public void after() {
+    // make sure after each test, the values of the ExpirationManager are reset
+    serverVM.invoke(ExpirableSecurityManager::reset);
+  }
+
+  @Test
+  public void currentClientShouldReAuthenticateWhenCredentialExpiredAndFunctionExecutionSucceed() {
+    clientShouldReAuthenticateWhenCredentialExpiredAndFunctionExecutionSucceed(
+        currentVersionClientVM);
+  }
+
+  @Test
+  public void releaseClientShouldReAuthenticateWhenCredentialExpiredAndFunctionExecutionSucceed() {
+    clientShouldReAuthenticateWhenCredentialExpiredAndFunctionExecutionSucceed(
+        releaseVersionClientVM);
+  }
+
+  private void clientShouldReAuthenticateWhenCredentialExpiredAndFunctionExecutionSucceed(
+      ClientVM clientVM) {
+    clientVM.invoke(() -> {
+      ClientCache clientCache = ClusterStartupRule.getClientCache();
+      assertThat(clientCache).isNotNull();
+      UpdatableUserAuthInitialize.setUser("data1");
+      ResultCollector rc = onServer(clientCache.getDefaultPool()).execute(writeFunction);
+      assertThat(((ArrayList) rc.getResult()).get(0))
+          .isEqualTo(TestFunctions.WriteFunction.SUCCESS_OUTPUT);
+    });
+
+    // expire the current user
+    serverVM.invoke(() -> ExpirableSecurityManager.addExpiredUser("data1"));
+
+    // do a second function execution, if this is successful, it means new credentials are provided
+    clientVM.invoke(() -> {
+      ClientCache clientCache = ClusterStartupRule.getClientCache();
+      assertThat(clientCache).isNotNull();
+      UpdatableUserAuthInitialize.setUser("data2");
+      ResultCollector rc = onServer(clientCache.getDefaultPool()).execute(writeFunction);
+      assertThat(((ArrayList) rc.getResult()).get(0))
+          .isEqualTo(TestFunctions.WriteFunction.SUCCESS_OUTPUT);
+    });
+
+    // all put operation succeeded
+    serverVM.invoke(() -> {
+      assertThat(ExpirableSecurityManager.getExpiredUsers().size()).isEqualTo(1);
```

Comment:
I would change this and the following size assertions to use:
```
assertThat(ExpirableSecurityManager.getExpiredUsers()).hasSize(1);
```
This will work as long as the subject (ie `getExpiredUsers()`) is a collection or array.

Any failure messages will then list all of the entries.

Reply:
I tried to use that method, but doesn't seem to work.

Reply:
I changed it to use that but had to add `.asList()`

Reply:
Didn't it complain about the object is not a list at runtime?

Reply:
> Didn't it complain about the object is not a list at runtime?

No, works fine.